### PR TITLE
6 create third lecture solution

### PR DESCRIPTION
**Pull request by grygy**

### CODE REVIEW

File: 03/passenger-service/src/test/java/cz/muni/fi/airportmanager/passengerservice/grpc/FlightCancellationServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+
+@QuarkusTest
+class FlightCancellationServiceTest {
+
+    @GrpcClient
+    MutinyFlightCancellationGrpc.MutinyFlightCancellationStub flightCancellationStub;
+
+    @Test
+    void testCancelFlight() {
+        var response = flightCancellationStub.cancelFlight(FlightCancellationRequest.newBuilder().setId(1).setReason("Unknown").build()).await().indefinitely();
```
Comment (grygy) [inline 03/passenger-service/src/test/java/cz/muni/fi/airportmanager/passengerservice/grpc/FlightCancellationServiceTest.java:20]: This cannot find test grpc service

**Review by xstefank — CHANGES_REQUESTED** [review]

File: 03/README.md [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+                license = @License(
+                        name = "Apache 2.0",
+                        url = "https://www.apache.org/licenses/LICENSE-2.0.html"))
+)
+public class FlightServiceApplication extends Application {
+    //...
+}
+```
+
+Note that `FlightServiceApplication` extends `Application` class from `jakarta.ws.rs.core.Application`. It's kinda "
```
[the remark this answers is not in the document]
Comment (grygy) [reply 03/README.md:87]: Used formal definition

File: 03/README.md [hunk 3]
```diff
@@ -0,0 +1,322 @@
...
+
+- HTTP/2 based - low latency, multiplexing, header compression, stream prioritization
+- High performance - binary protocol, efficient serialization, asynchronous by default
+- Bidirectional streaming - client and server can send messages at the same time
+
+### Differences between REST and gRPC
+
+| REST          | gRPC                         |
+|---------------|------------------------------|
+| HTTP/1.1      | HTTP/2                       |
```
[the remark this answers is not in the document]
Comment (grygy) [reply 03/README.md:174]: Yes true. Removed

File: 03/README.md [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+| Text based    | Binary                       |
+| JSON, XML     | Protocol Buffers             |
```
[the remark this answers is not in the document]
Comment (grygy) [reply 03/README.md:176]: Updated table

File: 03/README.md [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+| Stateless     | Bidirectional streaming      |
```
[the remark this answers is not in the document]
Comment (grygy) [reply 03/README.md:177]: Updated table

File: 03/README.md [hunk 6]
```diff
@@ -0,0 +1,322 @@
...
+| Client-server | Client-server, server-server |
```
[the remark this answers is not in the document]
Comment (grygy) [reply 03/README.md:178]: Table updated

File: 03/README.md [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+
+Now similar to injecting `FlightService` into `FlightResource` with `@Inject` we can inject `FlightCancellationService` into `FlightResource` with `@GrpcClient`. In `@GrpcClient` we need to provide the name of the service that we want to use. In our case it's `passenger-service`.
+
+The class of stub that we are injecting is `MutinyFlightCancellationGrpc.MutinyFlightCancellationStub`. Methods on this class are asynchronous and return `Uni` object. We can use `await().indefinitely()` to wait for the result.
+
+#### 4.2. Implement `cancelFlight` method in `FlightService`
+
+Implement `cancelFlight` method in `FlightService` that will call `cancelFlight` method on stub. You can use `await().indefinitely()` to wait for the result.
+
+#### 4.3. Test it
```
[the remark this answers is not in the document]
Comment (grygy) [reply 03/README.md:291]: Honestly, I don't know how to use the new dev UI to test grpc

image

File: 03/flight-service/src/main/resources/application.properties [hunk 8]
```diff
@@ -0,0 +1,7 @@
+quarkus.http.port=8079
+quarkus.http.test-port=8085
+
+quarkus.grpc.server.port=9000
+
+quarkus.grpc.clients.passenger-service.port=9002
+quarkus.grpc.clients.passenger-service.host=localhost
```
[the remark this answers is not in the document]
Comment (grygy) [reply 03/flight-service/src/main/resources/application.properties:7]: Doesn't look like url exists in `GrpcClientconfiguration`

[the remark this answers is not in the document]
Comment (xstefank) [reply 03/flight-service/src/main/resources/application.properties:7]: quarkus.grpc.clients.passenger-service.url?

**Review by xstefank — CHANGES_REQUESTED** [review]

**Comment by grygy** [comment]
moved to #20

File: 03/README.md [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+                license = @License(
+                        name = "Apache 2.0",
+                        url = "https://www.apache.org/licenses/LICENSE-2.0.html"))
+)
+public class FlightServiceApplication extends Application {
+    //...
+}
+```
+
+Note that `FlightServiceApplication` extends `Application` class from `jakarta.ws.rs.core.Application`. It defines the components of an application and supplies additional meta-data. It's not necessary to extend this class, but it's required for the use
```
[the remark this answers is not in the document]
Comment (grygy) [reply 03/README.md:87]: Changed

File: 03/flight-service/src/main/resources/application.properties (hunk 8 above)
[the remark this answers is not in the document]
Comment (grygy) [reply 03/flight-service/src/main/resources/application.properties:7]: Yes, there is no url.

File: 03/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (grygy) [reply 03/README.md:290]: When I moved proto files dev ui grpc works.

File: 03/README.md [hunk 10]
```diff
@@ -0,0 +1,321 @@
...
+
+#### 2.2. Generate classes
+
+Run `mvn compile` in project root to generate classes from `.proto` file.
+
+When you run this command, you should be able to generated classes in `flight-cancellation-api/target/generated-sources/grpc/cz/muni/fi/proto` directory. Check if they are there. This files will be used in `passenger-service` to implement gRPC service and in `flight-service` to implement gRPC stub.
+
+If you are using IntelliJ, run `mvn compile` from IDE Maven plugin under lifecycle. Idea has problem of recognizing generated classes. Or reload all maven projects. 
+
+It is also possible to run compile elsewhere and just mark the generated sources as sources dir by right click.
```
[the remark this answers is not in the document]
Comment (grygy) [reply 03/README.md:260]: Moved to the troubleshoot section because with the proto file moved to the service, it idea works as expected.

**Pull request closed without merging**